pageserver/src/layered_repository/delta_layer.rs

```diff
@@ -7,87 +7,164 @@
+//! "values". The summary is a fixed size header at the beginning of the file,
+//! and it contains basic information about the layer, and offsets to the other
+//! parts. The "index" is a B-tree, mapping from Key and LSN to an offset in the
+//! "values" part.  The actual page images and WAL records are stored in the
+//! "values" part.
 //!
 use crate::config::PageServerConf;
+use crate::layered_repository::blob_io::{BlobCursor, BlobWriter, WriteBlobWriter};
+use crate::layered_repository::block_io::{BlockBuf, BlockCursor, BlockReader, FileBlockReader};
+use crate::layered_repository::disk_btree::{DiskBtreeBuilder, DiskBtreeReader, VisitDirection};
 use crate::layered_repository::filename::{DeltaFileName, PathOrConf};
 use crate::layered_repository::storage_layer::{
-    BlobRef, Layer, ValueReconstructResult, ValueReconstructState,
+    Layer, ValueReconstructResult, ValueReconstructState,
 };
-use crate::repository::{Key, Value};
+use crate::page_cache::{PageReadGuard, PAGE_SZ};
+use crate::repository::{Key, Value, KEY_SIZE};
 use crate::virtual_file::VirtualFile;
 use crate::walrecord;
-use crate::DELTA_FILE_MAGIC;
 use crate::{ZTenantId, ZTimelineId};
-use anyhow::{bail, ensure, Result};
+use crate::{DELTA_FILE_MAGIC, STORAGE_FORMAT_VERSION};
+use anyhow::{bail, ensure, Context, Result};
 use log::*;
 use serde::{Deserialize, Serialize};
-use std::collections::HashMap;
-use zenith_utils::vec_map::VecMap;
 // avoid binding to Write (conflicts with std::io::Write)
 // while being able to use std::fmt::Write's methods
 use std::fmt::Write as _;
 use std::fs;
-use std::io::BufWriter;
-use std::io::Write;
+use std::io::{BufWriter, Write};
+use std::io::{Seek, SeekFrom};
 use std::ops::Range;
 use std::os::unix::fs::FileExt;
 use std::path::{Path, PathBuf};
-use std::sync::{RwLock, RwLockReadGuard, RwLockWriteGuard, TryLockError};
-
-use bookfile::{Book, BookWriter, ChapterWriter};
+use std::sync::{RwLock, RwLockReadGuard, RwLockWriteGuard};
 
 use zenith_utils::bin_ser::BeSer;
 use zenith_utils::lsn::Lsn;
 
-/// Mapping from (key, lsn) -> page/WAL record
-/// byte ranges in VALUES_CHAPTER
-static INDEX_CHAPTER: u64 = 1;
-
-/// Page/WAL bytes - cannot be interpreted
-/// without the page versions from the INDEX_CHAPTER
-static VALUES_CHAPTER: u64 = 2;
-
-/// Contains the [`Summary`] struct
-static SUMMARY_CHAPTER: u64 = 3;
-
+///
+/// Header stored in the beginning of the file
+///
+/// After this comes the 'values' part, starting on block 1. After that,
+/// the 'index' starts at the block indicated by 'index_start_blk'
+///
 #[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
 struct Summary {
+    /// Magic value to identify this as a zenith delta file. Always DELTA_FILE_MAGIC.
+    magic: u16,
+    format_version: u16,
+
     tenantid: ZTenantId,
     timelineid: ZTimelineId,
     key_range: Range<Key>,
     lsn_range: Range<Lsn>,
+
+    /// Block number where the 'index' part of the file begins.
+    index_start_blk: u32,
```

Comment:
Why do we need it? Why not to store index at the beginning of layer file?

Reply:
The way it currently works is that we build the index in-memory, as we write out all the WAL record/image blobs. After that, we write the index. You can't build the index before writing the data, because you don't know the offsets to store in the index, until you've written out the data.

We could mix the index pages with the blobs, by writing out each index page whenever it fills up. That would save some memory while the layer file is built. But I don't think it matters much.

We could also reserve the first block in the file for the index root page (after the summary), and build the rest of the index like we currently do. But it doesn't make any difference to performance.

Reply:
We can use two-phase algorithm: at first phase create index and at second - write records at correspondent offsets. It can be considered as unnecessary complication. But second stage can  be performed in parallel. I do not insist that writing index at the beginning of the file is much better than writing it at the end. Just some my thoughts. Also storing index of root pages seems to be not necessary - it can be always the first page of the index. Once again: doesn't seems to be critical, but allows to eliminate need to _load_ index. It may simplify code.